Add Relativistic Breit Wigner to TMath

JackHLindon opened the pull request:
# This Pull request:

## Changes or fixes:

TMath.cxx and TMath.h. 

Define BreitWignerRelativistic(...) in TMath.h and then calculate in TMath.cxx, following same convention for names of variables, default values and normalization as BreitWigner(...) which is already defined in TMath.

## Checklist:
- [x] tested changes locally

lmoneta reviewed (changes requested):
Hi, 
Thank you for the contribution. Could you please add also a test, for example using Google Test, in math/mathcore/test ? 
In addition it will be nice to have also a tutorial , in tutorials/math showing a plot with the difference between relativistic/non-relativistic Breit Wigner similar with what you have in the Forum post (https://root-forum.cern.ch/t/relativistic-breit-wigner-and-relativistic-voigt/48844/2)

JackHLindon commented:
Hi @lmoneta ,

Thank you for the help. I've looked at the other examples in tutorials/math and I gather the idea is just to produce a short stand alone script that shows how to use the function? I thought it would make sense to copy how it is done for the non relativistic BreitWigner function that already exists but there does not seem to be a tutorial for this so I made a tutorial file that covers both the non relativistic case that is already in ROOT and the relativistic case I am proposing. I have tried to follow the sort of way it is done in other tutorials.

(In particular I've copied the template comment at the top, though I'm not sure what the lines 
/// \file
/// \ingroup tutorial_math
/// \notebook
/// \macro_image
/// \macro_code
do. I assume this comment is read by some automatic documentation maker?)

For the test I apologise but I'm not too sure how to do this. I tried to see if there was a test already done for the nonrelativistic BreitWigner case that is already in ROOT I could compare to to try to understand what I need to do for this, but there does not seem to be. 

If in the math/mathcore/test directory I do `grep -R -i "breit" . ` , there are some results but they all seem to be testing the distribution version of the breitwigner in ProbFuncMathCore rather than the version defined in TMath. Would you be able to point me to an example of how to produce a test for this? (ideally an example that was done with the non relativistic version of the breit wigner that is already implemented if this test exists and I've just missed it, since the test for the relativistic version should be almost identical).

Thanks again,
Jack

JackHLindon commented:
Hi @lmoneta ,

I've added a function to testTMath.cxx, 

void testBreitWignerRelativistic()

I've tried to follow how the other tests are done but I am quite confused, they just seem to cout things not use any assertions. Is this correct? I have just couted 10 points across the relativistic breit wigner. Is there somewhere a test has been implemented for the non relativistic breit wigner?

Thanks,
Jack

lmoneta commented on math/mathcore/src/TMath.cxx:
```diff
@@ -440,6 +440,24 @@ Double_t TMath::BreitWigner(Double_t x, Double_t mean, Double_t gamma)
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Calculate a Relativistic Breit Wigner function with mean and gamma.
```
Can you add in the documentation the mathematical formula of the function in latex format. 
See example at https://github.com/root-project/root/blob/master/math/mathcore/inc/Math/SpecFuncMathCore.h#L43

lmoneta commented on math/mathcore/test/testTMath.cxx:
```diff
@@ -141,6 +141,20 @@ void testPlane()
         << endl;
 }
 
+void testBreitWignerRelativistic(){
```
The test is just printing the values. It would be nice to have a new executable and not included in testTMath where gtest is used. 
In this case we compare obtained values with some reference ones which can be hard-coded in the test. 
Here is an example: https://github.com/root-project/root/blob/master/math/mathcore/test/testKahan.cxx

lmoneta reviewed (commented):
I have some comments for the test and the documentation. 
In the case the test, if you don't have the time it could be done with another PR later. 

It would be also nice if you re-write the commits history in one or a couple of commits merging them.

JackHLindon commented:
Hi @lmoneta ,

I am very sorry for the long delay. 

I have added the mathematical formula of the function in latex format in the documentation as requested (it is quite a long expression fully written out).

In terms of "It would be also nice if you re-write the commits history in one or a couple of commits merging them."

I don't think I have the permissions to be able to do this in this pull request (if I do I cannot see how to do it). I can do it and then submit another pull request, or I think you have permissions to do it when accepting the merge by selectin by selecting "Squash and merge". In which case a commit message summarizing everything is:

"Define BreitWignerRelativistic(Double_t x, Double_t median, Double_t gamma) in TMath.h and then calculate in TMath.cxx, a similar non relativistic function BreitWigner already exists in the same location which was used as a template. A tutorial BreitWigner.C has been added in tutorials/math which produces plots comparing the non relativistic and relativistic case"

In terms of the test I apologise but I've looked at the example and I don't really understand how to implement it. Given that the BreitWigner function that already exists in TMath does not seem to have a test itself I would prefer to go ahead with potentially adding it later as you suggest.

Thanks for your help,
Jack

lmoneta commented:
Sorry for my late reply, I have missed your notification. 
Thank you for updating the documentation with the formula. 
For merging the commits you can do it locally in your branch, then force push in your GitHub and the PR will be automatically updated. 
For editing the commits you can use `git rebase -i`. See for example https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history or similar pages. 

There is also now a conflict in TMath, after some recent changes there. Could you also fix the conflicts by rebasing your branch to the master. Thanks

lmoneta commented:
Sorry for my late reply, I have missed your notification. 
Thank you for updating the documentation with the formula. 
For merging the commits you can do it locally in your branch, then force push in your GitHub and the PR will be automatically updated. 
For editing the commits you can use `git rebase -i`. See for example https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history or similar pages. 

There is also now a conflict in TMath, after some recent changes there. Could you also fix the conflicts by rebasing your branch to the master. 
Thank you for doing this!

JackHLindon commented:
Hi @lmoneta ,

I have squashed the commits and resolved the merge conflict merging with master, I think this is correct (I am unable to squash the merge with master and the previous commit...)

I noticed the merge conflict is because of changes to comments changing "calculate" to "calculates" and in the LaTeX comments changing "\f[" and "\f]" to "\f$" so I updated to be consistent with this. 

If this is ok I'll hopefully submit a pull request in the next few days to add the relativistic voigt as well as we discussed previously.

Thanks for your help,
Jack

lmoneta commented:
Thank you for update! 
I will test and if the result is good I will merge it in the master

lmoneta commented:
@JackHLindon : the builds fail because there is a problem with your second commit, a merge commit. Can you please just apply your first commit on top of the current master. 
If you cannot do on this current branch, please create a new branch on top of master and make a new PR

JackHLindon commented:
Hi @lmoneta ,

To make things easier as I was struggling I have got a fresh fork up to date with the master branch and put my changes in there in a new pull request: https://github.com/root-project/root/pull/10760 

Cheers,
Jack
